packages/quantic/force-app/main/default/lwc/quanticFacetManager/quanticFacetManager.js

```diff
@@ -0,0 +1,136 @@
+import { initializeWithHeadless, registerComponentForInit } from 'c/quanticHeadlessLoader';
+import { api, LightningElement, track } from 'lwc';
+
+/** @typedef {import("coveo").FacetManager} FacetManager */
+/** @typedef {import("coveo").FacetManagerState} FacetManagerState */
+/** @typedef {import("coveo").SearchStatus} SearchStatus */
+/** @typedef {import("coveo").SearchStatusState} SearchStatusState */
+/** @typedef {import("coveo").SearchEngine} SearchEngine */
+
+
+/**
+ * The `QuanticFacetManager` component acts as a container component allowing facets to be reordered dynamically as search queries are performed.
+ * 
+ * An item template element can be assigned to the `itemTemplate` slot allowing to customize the element that wraps each facet.
+ * 
+ * @example
+ * <c-quantic-facet-manager engine-id={engineId}>
+ *   <c-quantic-facet engine-id={engineId} field="type"></c-quantic-facet>
+ *   <c-quantic-facet engine-id={engineId} field="author"></c-quantic-facet>
+ * 
+ *   <div slot="itemTemplate" class="slds-var-m-bottom_large"></div>
+ * </c-quantic-facet-manager>
+ */
+export default class QuanticFacetManager extends LightningElement {
+  /**
+   * The ID of the engine instance the component registers to.
+   */
+  @api engineId;
+
+  /** @type {FacetManagerState} */
+  @track facetManagerState;
+  /** @type {SearchStatusState} */
+  @track searchStatusState;
+
+  /** @type {FacetManager} */
+  facetManager;
+  /** @type {Function} */
+  unsubscribeFacetManager;
+  /** @type {SearchStatus} */
+  searchStatus;
+  /** @type {Function} */
+  unsubscribeSearchStatus;
+
+  host;
+  itemTemplate;
+
+  connectedCallback() {
+    registerComponentForInit(this, this.engineId);
+  }
+
+  renderedCallback() {
+    initializeWithHeadless(this, this.engineId, this.initialize);
+
+    this.host = this.template.querySelector('.facet-manager__host');
+    this.resolveItemTemplate();
+    this.moveFacetsToHost();
+  }
+
+  disconnectedCallback() {
+    this.unsubscribeFacetManager?.();
+    this.unsubscribeSearchStatus?.();
+  }
+
+  /**
+   * @param {SearchEngine} engine 
+   */
+  initialize = (engine) => {
+    this.facetManager = CoveoHeadless.buildFacetManager(engine);
+    this.unsubscribeFacetManager = this.facetManager.subscribe(() => this.updateFacetManagerState());
+    this.searchStatus = CoveoHeadless.buildSearchStatus(engine);
+    this.unsubscribeSearchStatus = this.searchStatus.subscribe(() => this.updateSearchStatusState());
+  }
+
+  updateFacetManagerState() {
+    this.reorderFacets();
+  }
+
+  updateSearchStatusState() {
+    this.searchStatusState = this.searchStatus.state;
+  }
+
+  moveFacetsToHost() {
+    const facets = this.querySelectorAll('*');
+    facets.forEach((facet) => {
+      const wrapper = this.itemTemplate.cloneNode(false);
+      // @ts-ignore
+      wrapper.setAttribute('data-facet-id', facet.facetId ?? facet.field);
```

Comment:
I have same problem with typescript, I think we need to put time to figure out how to resolve it

Reply:
Ah i think this might work
```
/** @typedef {import('c/quanticFacet').default} quanticFacet */
```

```
/**
  * @returns {quanticFacet}
  */
get allFacets() {
  return this.querySelectorAll('*');
}
```
Have you tried something like this?

Reply:
Ah yes fair enough. you could typedef a new type within the facetManager too. or leave the @ts-ignore. Either way I'm okay with it personally. The type checking is something we added on top to help catch some errors but I don't think we should let it hinder us.